File: backend/app/schemas/application.py
"""
Pydantic schemas for application tracking
"""

from datetime import datetime, date
from typing import Optional
import uuid

from pydantic import BaseModel, Field


# ========== Application Schemas ==========

class JobApplicationBase(BaseModel):
    """Base application schema"""

    job_id: uuid.UUID = Field(..., description="Job ID")
    status: str = Field(default="applied", description="Application status")
    applied_via: Optional[str] = Field(None, description="Where application was submitted")
    cover_letter_id: Optional[uuid.UUID] = Field(None, description="Associated cover letter")
    resume_id: Optional[uuid.UUID] = Field(None, description="Associated resume")
    notes: Optional[str] = Field(None, description="Application notes")


class JobApplicationCreate(BaseModel):
    """Create application request"""

    job_id: uuid.UUID = Field(..., description="Job ID")
    applied_via: Optional[str] = Field(None, description="Where application was submitted")
    cover_letter_id: Optional[uuid.UUID] = Field(None, description="Associated cover letter")
    resume_id: Optional[uuid.UUID] = Field(None, description="Associated resume")
    notes: Optional[str] = Field(None, description="Application notes")


class JobApplicationUpdate(BaseModel):
    """Update application request"""

    status: Optional[str] = Field(None, description="New status")
    notes: Optional[str] = Field(None, description="Update notes")
    cover_letter_id: Optional[uuid.UUID] = Field(None, description="Update cover letter")
    resume_id: Optional[uuid.UUID] = Field(None, description="Update resume")


class JobApplicationResponse(BaseModel):
    """Application response"""

    id: uuid.UUID
    user_id: uuid.UUID
    job_id: uuid.UUID
    status: str
    application_date: datetime
    applied_via: Optional[str]
    cover_letter_id: Optional[uuid.UUID]
    resume_id: Optional[uuid.UUID]
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class JobApplicationDetailResponse(JobApplicationResponse):
    """Detailed application with interviews and offer"""

    interviews: list["InterviewResponse"] = Field(default_factory=list)
    activities: list["ApplicationActivityResponse"] = Field(default_factory=list)
    offer: Optional["JobOfferResponse"] = None

    class Config:
        from_attributes = True


# ========== Interview Schemas ==========

class InterviewBase(BaseModel):
    """Base interview schema"""

    interview_type: Optional[str] = Field(None, description="Interview type")
    scheduled_date: Optional[datetime] = Field(None, description="Scheduled date/time")
    duration_minutes: Optional[int] = Field(None, description="Duration in minutes")
    interviewer_name: Optional[str] = Field(None, description="Interviewer name")
    interviewer_email: Optional[str] = Field(None, description="Interviewer email")
    meeting_link: Optional[str] = Field(None, description="Meeting link")
    preparation_notes: Optional[str] = Field(None, description="Preparation notes")


class InterviewCreate(InterviewBase):
    """Create interview request"""

    pass


class InterviewUpdate(BaseModel):
    """Update interview request"""

    interview_type: Optional[str] = Field(None)
    scheduled_date: Optional[datetime] = Field(None)
    duration_minutes: Optional[int] = Field(None)
    interviewer_name: Optional[str] = Field(None)
    interviewer_email: Optional[str] = Field(None)
    meeting_link: Optional[str] = Field(None)
    preparation_notes: Optional[str] = Field(None)
    status: Optional[str] = Field(None, description="Interview status")
    feedback: Optional[str] = Field(None, description="Feedback")
    interview_score: Optional[int] = Field(None, ge=1, le=10, description="Score 1-10")


class InterviewResponse(BaseModel):
    """Interview response"""

    id: uuid.UUID
    application_id: uuid.UUID
    interview_type: Optional[str]
    scheduled_date: Optional[datetime]
    duration_minutes: Optional[int]
    interviewer_name: Optional[str]
    interviewer_email: Optional[str]
    meeting_link: Optional[str]
    preparation_notes: Optional[str]
    feedback: Optional[str]
    interview_score: Optional[int]
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ========== Activity Schemas ==========

class ApplicationActivityResponse(BaseModel):
    """Activity log response"""

    id: uuid.UUID
    application_id: uuid.UUID
    activity_type: str
    description: Optional[str]
    previous_status: Optional[str]
    new_status: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# ========== Offer Schemas ==========

class JobOfferBase(BaseModel):
    """Base offer schema"""

    salary: Optional[int] = Field(None, description="Annual salary")
    start_date: Optional[date] = Field(None, description="Start date")
    bonus: Optional[int] = Field(None, description="Bonus amount")
    benefits: Optional[str] = Field(None, description="Benefits description")
    offer_letter_url: Optional[str] = Field(None, description="Offer letter URL")
    offer_expiration_date: Optional[date] = Field(None, description="Expiration date")


class JobOfferCreate(JobOfferBase):
    """Create offer request"""

    pass


class JobOfferUpdate(BaseModel):
    """Update offer request"""

    status: Optional[str] = Field(None, description="Offer status")
    salary: Optional[int] = Field(None)
    start_date: Optional[date] = Field(None)
    bonus: Optional[int] = Field(None)
    benefits: Optional[str] = Field(None)
    offer_letter_url: Optional[str] = Field(None)
    offer_expiration_date: Optional[date] = Field(None)
    negotiation_notes: Optional[str] = Field(None, description="Negotiation notes")


class JobOfferResponse(BaseModel):
    """Offer response"""

    id: uuid.UUID
    application_id: uuid.UUID
    status: str
    salary: Optional[int]
    start_date: Optional[date]
    bonus: Optional[int]
    benefits: Optional[str]
    offer_letter_url: Optional[str]
    offer_expiration_date: Optional[date]
    negotiation_notes: Optional[str]
    accepted_date: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ========== Bulk Operations ==========

class BulkStatusUpdateRequest(BaseModel):
    """Bulk update status request"""

    application_ids: list[uuid.UUID] = Field(..., description="List of application IDs")
    new_status: str = Field(..., description="New status for all applications")
    notes: Optional[str] = Field(None, description="Notes for status change")


class BulkStatusUpdateResponse(BaseModel):
    """Bulk update response"""

    updated: int = Field(..., description="Number of applications updated")
    failed: int = Field(default=0, description="Number that failed")
    status: str


# ========== Dashboard/Analytics ==========

class ApplicationStats(BaseModel):
    """Application statistics"""

    total_applications: int
    status_breakdown: dict[str, int]  # {status: count}
    total_interviews: int
    completed_interviews: int
    offers_received: int
    offers_accepted: int
    rejection_rate: float
    average_time_to_interview: Optional[int]  # days
    average_time_to_offer: Optional[int]  # days


class ApplicationSummary(BaseModel):
    """Summary of application activity"""

    application_id: uuid.UUID
    job_title: str
    company: str
    status: str
    days_in_status: int
    next_step: Optional[str]
    interview_count: int
    last_activity: Optional[str]
    last_activity_date: Optional[datetime]

File: frontend/src/lib/types/application.ts
/**
 * Job Application Types
 */

export interface JobApplication {
  id: string;
  user_id: string;
  job_id: string;
  status: ApplicationStatus;
  application_date: string;
  applied_via?: string;
  cover_letter_id?: string;
  resume_id?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface JobApplicationDetail extends JobApplication {
  interviews: Interview[];
  activities: ApplicationActivity[];
  offer?: JobOffer;
}

export type ApplicationStatus =
  | "applied"
  | "under_review"
  | "interview_scheduled"
  | "interviewed"
  | "offer_received"
  | "rejected"
  | "withdrawn"
  | "accepted";

export interface Interview {
  id: string;
  application_id: string;
  interview_type?: string;
  scheduled_date?: string;
  duration_minutes?: number;
  interviewer_name?: string;
  interviewer_email?: string;
  meeting_link?: string;
  preparation_notes?: string;
  feedback?: string;
  interview_score?: number;
  status: InterviewStatus;
  created_at: string;
  updated_at: string;
}

export type InterviewStatus = "scheduled" | "completed" | "cancelled" | "rescheduled";

export interface ApplicationActivity {
  id: string;
  application_id: string;
  activity_type: string;
  description?: string;
  previous_status?: string;
  new_status?: string;
  created_at: string;
}

export interface JobOffer {
  id: string;
  application_id: string;
  status: OfferStatus;
  salary?: number;
  start_date?: string;
  bonus?: number;
  benefits?: string;
  offer_letter_url?: string;
  offer_expiration_date?: string;
  negotiation_notes?: string;
  accepted_date?: string;
  created_at: string;
  updated_at: string;
}

export type OfferStatus = "received" | "accepted" | "declined" | "pending";

// Request/Response types

export interface JobApplicationCreateRequest {
  job_id: string;
  applied_via?: string;
  cover_letter_id?: string;
  resume_id?: string;
  notes?: string;
}

export interface JobApplicationUpdateRequest {
  status?: ApplicationStatus;
  notes?: string;
  cover_letter_id?: string;
  resume_id?: string;
}

export interface InterviewCreateRequest {
  interview_type?: string;
  scheduled_date?: string;
  duration_minutes?: number;
  interviewer_name?: string;
  interviewer_email?: string;
  meeting_link?: string;
  preparation_notes?: string;
}

export interface InterviewUpdateRequest extends InterviewCreateRequest {
  status?: InterviewStatus;
  feedback?: string;
  interview_score?: number;
}

export interface JobOfferCreateRequest {
  salary?: number;
  start_date?: string;
  bonus?: number;
  benefits?: string;
  offer_letter_url?: string;
  offer_expiration_date?: string;
}

export interface JobOfferUpdateRequest extends JobOfferCreateRequest {
  status?: OfferStatus;
  negotiation_notes?: string;
}

export interface BulkStatusUpdateRequest {
  application_ids: string[];
  new_status: ApplicationStatus;
  notes?: string;
}

export interface ApplicationStats {
  total_applications: number;
  status_breakdown: Record<ApplicationStatus, number>;
  total_interviews: number;
  completed_interviews: number;
  offers_received: number;
  offers_accepted: number;
  rejection_rate: number;
  average_time_to_interview?: number;
  average_time_to_offer?: number;
}

export interface ApplicationSummary {
  application_id: string;
  job_title: string;
  company: string;
  status: ApplicationStatus;
  days_in_status: number;
  next_step?: string;
  interview_count: number;
  last_activity?: string;
  last_activity_date?: string;
}
